docs/reference/scripting/security.asciidoc

```diff
@@ -1,84 +1,31 @@
-allow any successful effort to circumvent the other security layers to do
-*anything* on your server. Elasticsearch will refuse to start if it detects
-that it is running as `root` but this is so important that it is worth double
-and triple checking.
-
-[discrete]
-=== Do not expose Elasticsearch directly to users
-Do not expose Elasticsearch directly to users, instead have an application
-make requests on behalf of users. If this is not possible, have an application
-to sanitize requests from users. If *that* is not possible then have some
-mechanism to track which users did what. Understand that it is quite possible
-to write a <<search, `_search`>> that overwhelms Elasticsearch and brings down
-the cluster. All such searches should be considered bugs and the Elasticsearch
-contributors make an effort to prevent this but they are still possible.
-
-[discrete]
-=== Do not expose Elasticsearch directly to the Internet
-Do not expose Elasticsearch to the Internet, instead have an application
-make requests on behalf of the Internet. Do not entertain the thought of having
-an application "sanitize" requests to Elasticsearch. Understand that it is
-possible for a sufficiently determined malicious user to write searches that
-overwhelm the Elasticsearch cluster and bring it down. For example:
-
-Good:
-
-* Users type text into a search box and the text is sent directly to a
-<<query-dsl-match-query>>, <<query-dsl-match-query-phrase>>,
-<<query-dsl-simple-query-string-query>>, or any of the <<search-suggesters>>.
-* Running a script with any of the above queries that was written as part of
-the application development process.
-* Running a script with `params` provided by users.
-* User actions makes documents with a fixed structure.
-
-Bad:
-
-* Users can write arbitrary scripts, queries, `_search` requests.
-* User actions make documents with structure defined by users.
-
-[discrete]
-[[modules-scripting-other-layers]]
-=== Other security layers
-In addition to user privileges and script sandboxing Elasticsearch uses the
-https://www.oracle.com/java/technologies/javase/seccodeguide.html[Java Security Manager]
+In addition to user privileges and script sandboxing, {es} uses the
+https://www.oracle.com/java/technologies/javase/seccodeguide.html[Java Security Manager (JVM)]
 and native security tools as additional layers of security.
 
-As part of its startup sequence Elasticsearch enables the Java Security Manager
-which limits the actions that can be taken by portions of the code. Painless
-uses this to limit the actions that generated Painless scripts can take,
-preventing them from being able to do things like write files and listen to
-sockets.
+As part of its startup sequence, {es} enables the JVM to limit the actions that
+can be taken by portions of the code. Painless uses this to limit the actions
+that generated Painless scripts can take, preventing them from being able to do 
+things like write files and listen to sockets.
```

Comment:
Is this an accurate statement?
>Painless uses the Java Security Manager as a layer in a defense in depth strategy to prevent scripts from doing things like writing files and listening to sockets.

Reply:
Yes.

Interesting technical details if you care to read them: Painless emits scripts as few [privileges](https://github.com/elastic/elasticsearch/blob/master/modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java#L51-L66) as possible. That's pretty much all it has to do here. ES has to go through some hoops when scripts make calls that have to run with more privileges like loading doc values. But ES it is responsible for escalating the privileges and letting go. It took us a while to get it all in order but in the end its surprisingly simple. Seriously, you stare at it for an hour or too and it comes together and you think you know everything.

Reply:
Cool! Maybe it's worth expanding the next section as follows: 

> The following sections describe security settings for scripts and how to alter the default settings. Take extreme caution when providing greater permissions to scripts beyond the default settings. When running a script, Painless uses the least privileges possible. Any extra permissions that you allow weakens the total security of your {es} deployment.

Reply:
A bit of nuance that might not matter: the java security manager applies to all code. It's unique in that we can lower privileges for some code and raise it for others. Which is an interesting story for scripts.

seccomp/seatbelt/ActiveProcessLimit is application wide. We enable them very very very early in startup and drop most of our privileges. We can't get those privileges back. We can't lower the privileges for specific parts of the code either. So its another layer, but its not quite specific to scripting.


It might be educational to say which layers prevent a script from dumping the file containing hashed password like:
```
Files.readAllLines('/etc/shadow').forEach(this::emit);
```


Painless doesn't expose this method to scripts at all. It doesn't expose any methods to read files. The script won't even compile.

You could try subverting the sandbox with reflection like you could back when groovy was our scripting language. Something like:
```
Class.forName("java.nio.file.Files").getMethod("readAllLines").invoke(null, new Object[] {'/etc/shadow'/}).forEach(this::emit);
```
But that'd fail because we check the whitelist during reflection.

But we're not perfect. Sure, we have tests and we're never going to release a version of ES without this protection but folks are tricky. Say you could socially engineer folks into installing an ES plugin that extends the scripting API to expose this. That'd be possible - remember ES can expand its privileges when its taking action for a script. It has to be able to open the files with doc values in them, for example. So the plugin could expose something like:
```
iPromiseThisIsntReadingYouPasswords().forEach(this::emit);
```

Here we have a few more defenses. When you install this funny plugin you'll see a list of privileges that it wants - including that it wants to be able to read from weird places on the filesystem. You have to look at this list and approve it. But folks often just say "yes" or then run the plugin installed unattended (terraform, docker setup, etc) so they miss this.

So that warning is fairly thin. But it still won't work! When the plugin went to go read from `/etc/shadow` with it's escalated privileges it'd fail because we use seccomp to drop our permission to read from most paths on the filesystem on startup. It'd get some kind of weird error message. Seccomp gives kind of garbage error messages. It doesn't say "seccomp is blocking you".

Ok. So what if you disable seccomp. You can turn it off in 7.x I believe. Hell, you can comment it out and recompile ES if you really want. Well, you *still* can't read `/etc/shadow` because you aren't running as root. But if you can run arbitrary code you could probably find a way around it. Bugs are bugs.

So keep your kernel patched so folks can't use these bugs. Keep docker up to date. That thing has had some bugs.

Remember how I said you could just comment things out? That's why you need to check the hash and signature. Remember the SolarWinds hack? We're careful with the libraries we pull in. You should be careful you get what we publish.

None of this is perfect. But there are layers and layers and layers to hopefully stop folks. Don't disable any of the layers. Don't make it easy.

Worse, we're just one access point. We try not to be the weak one.

Reply:
I'm probably forgetting some layers too, frankly.

Reply:
It's probably also worth saying why we keep talking about scripting. Security matters outside of scripting, but scripting is a very easy potential point of attack. Back in the bad old days when we ran mvel the only protection we had against all this was "don't run as root". Which is weak - like I said, OSes have bugs. And we didn't used to refuse to start when you run us as root. Plenty of folks were upset for that change.

To someone who thinks about security "I'll let anyone that can send me an HTTP request write some code and run it for them" sounds like a disaster. Even if you have to have a user name and password on the request. I've had security folks tell me that to my face. Back then I was younger and stupider and said something like "yeah, we make sure only good folks can get on the network that can send HTTP requests". They spent the next hour teaching me about security in depth. I didn't work on ES at the time, but the folks that did disabled scripts sent over the wire shortly after.

All of these layers help and we're always excited to have more.